File: api/check_spam.py
import os
import json
from twilio.rest import Client
from twilio.base.exceptions import TwilioRestException

def handler(event, context):
    # Enable CORS
    headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type',
        'Access-Control-Allow-Methods': 'POST, OPTIONS'
    }

    # Handle preflight OPTIONS request
    if event['httpMethod'] == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': ''
        }

    try:
        # Parse the incoming request body
        body = json.loads(event['body'])
        phone_number = body.get('phoneNumber')
        secondary_phone_number = body.get('secondaryPhoneNumber')

        if not phone_number:
            return {
                'statusCode': 400,
                'headers': headers,
                'body': json.dumps({'error': 'Phone number is required'})
            }

        # Initialize Twilio client with environment variables
        account_sid = os.environ.get('TWILIO_ACCOUNT_SID', 'AC1122b3207e35e21162e277551dfd5191')
        auth_token = os.environ.get('TWILIO_AUTH_TOKEN', '389de81902c5b29cf6d9e0b506d08d84')
        nomorobo_addon_sid = os.environ.get('NOMOROBO_ADDON_SID', 'XE3e2d5ca759ed15c838f7b39fd0b81346')

        if not account_sid or not auth_token:
            return {
                'statusCode': 500,
                'headers': headers,
                'body': json.dumps({'error': 'Server configuration error'})
            }

        client = Client(account_sid, auth_token)

        # Prepare options for Twilio Lookup API
        lookup_options = {
            'add_ons': nomorobo_addon_sid
        }

        # Add secondary_address if provided
        if secondary_phone_number:
            # Format the parameter name using the Add-on SID
            param_name = f'add_ons.{nomorobo_addon_sid}.secondary_address'
            lookup_options[param_name] = secondary_phone_number

        try:
            # Call Twilio Lookup API with Nomorobo Spam Score Add-on
            phone_number_lookup = client.lookups.v1.phone_numbers(phone_number).fetch(**lookup_options)

            # Process the response
            # Get the results using the Add-on SID
            nomorobo_result = phone_number_lookup.add_ons.get('results', {}).get(nomorobo_addon_sid)

            if not nomorobo_result or nomorobo_result.get('status') != 'successful':
                return {
                    'statusCode': 500,
                    'headers': headers,
                    'body': json.dumps({
                        'error': 'Failed to retrieve spam score',
                        'details': nomorobo_result or 'No result from Nomorobo'
                    })
                }

            # Get the spam score from the result
            spam_score = nomorobo_result.get('result', {}).get('score')

            # Map the Nomorobo score to our application's format
            status = 'clean'
            risk_score = 0

            if spam_score == 1:
                status = 'flagged'
                risk_score = 85  # High risk score for confirmed spam
            else:
                # You could implement additional logic here based on other factors
                # For example, checking carrier information or call patterns
                status = 'clean'
                risk_score = 10  # Low risk score for non-spam

            # Return the formatted result
            return {
                'statusCode': 200,
                'headers': headers,
                'body': json.dumps({
                    'status': status,
                    'riskScore': risk_score,
                    'carriers': [
                        {'name': 'AT&T', 'status': 'flagged' if spam_score == 1 else 'clean'},
                        {'name': 'Verizon', 'status': 'flagged' if spam_score == 1 else 'clean'},
                        {'name': 'T-Mobile', 'status': 'flagged' if spam_score == 1 else 'clean'}
                    ],
                    'timeChecked': phone_number_lookup.date_created.isoformat(),
                    'recommendations': [
                        'Immediate number remediation required',
                        'Implement branded caller ID solutions',
                        'Register with carrier reputation systems',
                        'Consider number rotation strategy'
                    ] if spam_score == 1 else [
                        'Continue monitoring your number',
                        'Use branded caller ID to increase answer rates',
                        'Maintain consistent calling patterns'
                    ],
                    'rawData': {
                        'nomoroboScore': spam_score
                    }
                })
            }
        except TwilioRestException as e:
            return {
                'statusCode': 400,
                'headers': headers,
                'body': json.dumps({
                    'error': 'Failed to check phone number',
                    'message': str(e)
                })
            }

    except Exception as e:
        print(f'Error processing request: {str(e)}')

        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({
                'error': 'Internal server error',
                'message': str(e)
            })
        }

File: api/check-spam.js
// This file would be deployed as a serverless function
// For example, as a Netlify function, Vercel serverless function, or AWS Lambda

const twilio = require('twilio');

exports.handler = async function(event, context) {
  // Enable CORS
  const headers = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Access-Control-Allow-Methods': 'POST, OPTIONS'
  };

  // Handle preflight OPTIONS request
  if (event.httpMethod === 'OPTIONS') {
    return {
      statusCode: 200,
      headers,
      body: ''
    };
  }

  try {
    // Parse the incoming request body
    const data = JSON.parse(event.body);
    const { phoneNumber, secondaryPhoneNumber } = data;

    if (!phoneNumber) {
      return {
        statusCode: 400,
        headers,
        body: JSON.stringify({ error: 'Phone number is required' })
      };
    }

    // Initialize Twilio client with environment variables
    const accountSid = process.env.TWILIO_ACCOUNT_SID;
    const authToken = process.env.TWILIO_AUTH_TOKEN;
    const nomoroboAddOnSid = process.env.NOMOROBO_ADDON_SID;

    if (!accountSid || !authToken) {
      return {
        statusCode: 500,
        headers,
        body: JSON.stringify({ error: 'Server configuration error' })
      };
    }

    const client = twilio(accountSid, authToken);

    // Prepare options for Twilio Lookup API
    const lookupOptions = {
      addOns: nomoroboAddOnSid
    };

    // Add secondary_address if provided
    if (secondaryPhoneNumber) {
      // Format the parameter name using the Add-on SID
      const paramName = `addOns.${nomoroboAddOnSid}.secondary_address`;
      lookupOptions[paramName] = secondaryPhoneNumber;
    }

    // Call Twilio Lookup API with Nomorobo Spam Score Add-on
    const phoneNumberLookup = await client.lookups.v1
      .phoneNumbers(phoneNumber)
      .fetch(lookupOptions);

    // Process the response
    // Get the results using the Add-on SID
    const nomoroboResult = phoneNumberLookup.addOns?.results?.[nomoroboAddOnSid];

    if (!nomoroboResult || nomoroboResult.status !== 'successful') {
      return {
        statusCode: 500,
        headers,
        body: JSON.stringify({
          error: 'Failed to retrieve spam score',
          details: nomoroboResult || 'No result from Nomorobo'
        })
      };
    }

    const spamScore = nomoroboResult.result?.score;

    // Map the Nomorobo score to our application's format
    let status = 'clean';
    let riskScore = 0;

    if (spamScore === 1) {
      status = 'flagged';
      riskScore = 85; // High risk score for confirmed spam
    } else {
      // You could implement additional logic here based on other factors
      // For example, checking carrier information or call patterns
      status = 'clean';
      riskScore = 10; // Low risk score for non-spam
    }

    // Return the formatted result
    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({
        status,
        riskScore,
        carriers: [
          { name: 'AT&T', status: spamScore === 1 ? 'flagged' : 'clean' },
          { name: 'Verizon', status: spamScore === 1 ? 'flagged' : 'clean' },
          { name: 'T-Mobile', status: spamScore === 1 ? 'flagged' : 'clean' }
        ],
        timeChecked: new Date().toISOString(),
        recommendations: spamScore === 1 ? [
          'Immediate number remediation required',
          'Implement branded caller ID solutions',
          'Register with carrier reputation systems',
          'Consider number rotation strategy'
        ] : [
          'Continue monitoring your number',
          'Use branded caller ID to increase answer rates',
          'Maintain consistent calling patterns'
        ],
        rawData: {
          nomoroboScore: spamScore
        }
      })
    };
  } catch (error) {
    console.error('Error processing request:', error);

    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({
        error: 'Internal server error',
        message: error.message
      })
    };
  }
};
